File: football/Services/RateLimitManager.swift
import Foundation

/// API 레이트 리밋 매니저
@MainActor
class RateLimitManager {
    static let shared = RateLimitManager()
    
    /// 요청 기록
    private struct RequestRecord {
        let timestamp: Date
        let endpoint: String
    }
    
    private var requests: [RequestRecord] = []
    private var blocked = false
    private var blockUntil: Date?
    private var queue: [(CheckedContinuation<Void, Never>)] = []
    
    // 설정 - Rapid API 제한에 맞게 조정 (450 requests/minute)
    private let maxRequests = 400  // 450보다 낮게 설정하여 여유 확보
    private let windowInterval: TimeInterval = 60 // 1분
    private let retryAfter: TimeInterval = 65 // 65초 (1분 + 여유)
    
    private init() {
        // 주기적으로 오래된 요청 정리
        Timer.scheduledTimer(withTimeInterval: 10, repeats: true) { _ in
            Task { @MainActor in
                self.cleanOldRequests()
            }
        }
    }
    
    /// 요청 가능 여부 확인
    func canMakeRequest() -> Bool {
        cleanOldRequests()
        
        // 차단 상태 확인
        if blocked, let blockUntil = blockUntil {
            if Date() < blockUntil {
                return false
            } else {
                // 차단 해제
                self.blocked = false
                self.blockUntil = nil
                processQueue()
            }
        }
        
        return requests.count < maxRequests
    }
    
    /// 요청 기록
    func recordRequest(endpoint: String) {
        requests.append(RequestRecord(timestamp: Date(), endpoint: endpoint))
    }
    
    /// 요청 슬롯 대기
    func waitForSlot() async {
        if canMakeRequest() {
            return
        }
        
        await withCheckedContinuation { continuation in
            queue.append(continuation)
        }
    }
    
    /// 레이트 리밋 에러 처리
    func handleRateLimitError(retryAfter: TimeInterval? = nil) {
        blocked = true
        blockUntil = Date().addingTimeInterval(retryAfter ?? self.retryAfter)
        
        print("⚠️ Rate limit hit. Blocked until \(blockUntil!)")
    }
    
    /// 현재 상태
    func getStatus() -> (current: Int, max: Int, remaining: Int, resetTime: Date) {
        cleanOldRequests()
        
        return (
            current: requests.count,
            max: maxRequests,
            remaining: max(0, maxRequests - requests.count),
            resetTime: Date().addingTimeInterval(windowInterval)
        )
    }
    
    /// 엔드포인트별 통계
    func getEndpointStats() -> [String: Int] {
        cleanOldRequests()
        
        var stats: [String: Int] = [:]
        requests.forEach { record in
            stats[record.endpoint, default: 0] += 1
        }
        
        return stats
    }
    
    // MARK: - Private
    
    /// 오래된 요청 정리
    private func cleanOldRequests() {
        let cutoff = Date().addingTimeInterval(-windowInterval)
        requests.removeAll { $0.timestamp < cutoff }
    }
    
    /// 대기열 처리
    private func processQueue() {
        while !queue.isEmpty && canMakeRequest() {
            let continuation = queue.removeFirst()
            continuation.resume()
        }
    }
}

/// 레이트 리밋을 적용한 API 요청
func withRateLimit<T>(
    endpoint: String,
    request: () async throws -> T
) async throws -> T {
    // 요청 가능할 때까지 대기
    await RateLimitManager.shared.waitForSlot()
    
    // 요청 기록
    await MainActor.run {
        RateLimitManager.shared.recordRequest(endpoint: endpoint)
    }
    
    do {
        let result = try await request()
        return result
    } catch {
        // 429 에러 처리
        if let urlError = error as? URLError,
           urlError.code == .userAuthenticationRequired { // 429 상태 코드
            await MainActor.run {
                RateLimitManager.shared.handleRateLimitError()
            }
            throw FootballAPIError.rateLimitExceeded
        }
        
        throw error
    }
}

File: web/lib/utils/rate-limit-manager.ts
interface RateLimitConfig {
  maxRequests: number
  windowMs: number
  retryAfter?: number
}

interface RequestRecord {
  timestamp: number
  endpoint: string
}

class RateLimitManager {
  private requests: RequestRecord[] = []
  private queued: Array<() => void> = []
  private blocked = false
  private blockUntil = 0
  
  // 기본 설정: 분당 100개 요청
  private config: RateLimitConfig = {
    maxRequests: 100,
    windowMs: 60 * 1000, // 1분
    retryAfter: 5000 // 5초 후 재시도
  }
  
  // 설정 업데이트
  updateConfig(config: Partial<RateLimitConfig>): void {
    this.config = { ...this.config, ...config }
  }
  
  // 요청 가능 여부 확인
  canMakeRequest(): boolean {
    this.cleanOldRequests()
    
    // 차단 상태 확인
    if (this.blocked && Date.now() < this.blockUntil) {
      return false
    }
    
    // 차단 해제
    if (this.blocked && Date.now() >= this.blockUntil) {
      this.blocked = false
      this.processQueue()
    }
    
    return this.requests.length < this.config.maxRequests
  }
  
  // 요청 기록
  recordRequest(endpoint: string): void {
    this.requests.push({
      timestamp: Date.now(),
      endpoint
    })
  }
  
  // 오래된 요청 정리
  private cleanOldRequests(): void {
    const cutoff = Date.now() - this.config.windowMs
    this.requests = this.requests.filter(req => req.timestamp > cutoff)
  }
  
  // 대기열에 추가
  async waitForSlot(): Promise<void> {
    if (this.canMakeRequest()) {
      return
    }
    
    return new Promise(resolve => {
      this.queued.push(resolve)
    })
  }
  
  // 대기열 처리
  private processQueue(): void {
    while (this.queued.length > 0 && this.canMakeRequest()) {
      const resolve = this.queued.shift()
      if (resolve) resolve()
    }
  }
  
  // 레이트 리밋 에러 처리
  handleRateLimitError(retryAfter?: number): void {
    this.blocked = true
    this.blockUntil = Date.now() + (retryAfter || this.config.retryAfter!)
    
    console.warn(`⚠️ Rate limit hit. Blocked until ${new Date(this.blockUntil).toLocaleTimeString()}`)
  }
  
  // 현재 상태 가져오기
  getStatus(): {
    currentRequests: number
    maxRequests: number
    remainingRequests: number
    resetTime: Date
    blocked: boolean
  } {
    this.cleanOldRequests()
    
    return {
      currentRequests: this.requests.length,
      maxRequests: this.config.maxRequests,
      remainingRequests: Math.max(0, this.config.maxRequests - this.requests.length),
      resetTime: new Date(Date.now() + this.config.windowMs),
      blocked: this.blocked
    }
  }
  
  // 요청별 통계
  getEndpointStats(): Record<string, number> {
    this.cleanOldRequests()
    
    const stats: Record<string, number> = {}
    this.requests.forEach(req => {
      stats[req.endpoint] = (stats[req.endpoint] || 0) + 1
    })
    
    return stats
  }
}

// 싱글톤 인스턴스
export const rateLimiter = new RateLimitManager()

// API 요청 래퍼
export async function withRateLimit<T>(
  request: () => Promise<T>,
  endpoint: string
): Promise<T> {
  // 요청 가능할 때까지 대기
  await rateLimiter.waitForSlot()
  
  // 요청 기록
  rateLimiter.recordRequest(endpoint)
  
  try {
    const result = await request()
    return result
  } catch (error: any) {
    // 429 에러 처리
    if (error.status === 429) {
      const retryAfter = error.headers?.['retry-after']
      const retryMs = retryAfter ? parseInt(retryAfter) * 1000 : undefined
      rateLimiter.handleRateLimitError(retryMs)
      
      // 에러 재발생
      throw new Error(`Rate limit exceeded. Retry after ${retryAfter} seconds`)
    }
    
    throw error
  }
}

// 배치 요청 관리
export class BatchRequestManager {
  private batch: Array<{
    request: () => Promise<any>
    resolve: (value: any) => void
    reject: (error: any) => void
  }> = []
  
  private batchSize = 5
  private batchDelay = 1000 // 1초
  private processing = false
  
  // 배치에 요청 추가
  add<T>(request: () => Promise<T>): Promise<T> {
    return new Promise((resolve, reject) => {
      this.batch.push({ request, resolve, reject })
      
      if (!this.processing) {
        this.processBatch()
      }
    })
  }
  
  // 배치 처리
  private async processBatch(): Promise<void> {
    if (this.processing || this.batch.length === 0) return
    
    this.processing = true
    
    while (this.batch.length > 0) {
      // 배치 크기만큼 가져오기
      const currentBatch = this.batch.splice(0, this.batchSize)
      
      // 병렬 처리
      const promises = currentBatch.map(async ({ request, resolve, reject }) => {
        try {
          const result = await withRateLimit(request, 'batch')
          resolve(result)
        } catch (error) {
          reject(error)
        }
      })
      
      await Promise.allSettled(promises)
      
      // 다음 배치 전 대기
      if (this.batch.length > 0) {
        await new Promise(resolve => setTimeout(resolve, this.batchDelay))
      }
    }
    
    this.processing = false
  }
}

export const batchManager = new BatchRequestManager()